diff --git a/player/screenshot.c b/player/screenshot.c
index c733a2bdb14909b24d476f2e277de0f5b4ef5def..06298c1957f8d9e5f97c75138c1de3344295b20f 100644
--- a/player/screenshot.c
+++ b/player/screenshot.c
@@ -377,7 +377,8 @@ static struct mp_image *screenshot_get(struct MPContext *mpctx, int mode,
     }
 
     // vo_get_current_frame() can return a hardware frame, which we have to download first.
-    if (image && image->fmt.flags & MP_IMGFLAG_HWACCEL) {
+    if (image && (IMGFMT_IS_HWACCEL(image->imgfmt) ||
+                  (image->fmt.flags & MP_IMGFLAG_HWACCEL))) {
         struct mp_image *nimage = mp_image_hw_download(image, NULL);
         talloc_free(image);
         if (!nimage)
diff --git a/video/image_writer.c b/video/image_writer.c
index f3c9777026c6082fa86a572838fc6f58687f7c92..592b6d7d6aa0b59335deaf3d3a7d1c0af71bdedb 100644
--- a/video/image_writer.c
+++ b/video/image_writer.c
@@ -46,6 +46,7 @@
 #include "video/fmt-conversion.h"
 #include "video/img_format.h"
 #include "video/mp_image.h"
+#include "video/mp_image_pool.h"
 #include "video/sws_utils.h"
 
 #include "options/m_option.h"
@@ -599,8 +600,17 @@ static struct mp_image *convert_image(struct mp_image *image, int destfmt,
                                       struct mpv_global *global,
                                       struct mp_log *log)
 {
+    struct mp_image *base = image;
+    if (IMGFMT_IS_HWACCEL(base->imgfmt) || (base->fmt.flags & MP_IMGFLAG_HWACCEL)) {
+        base = mp_image_hw_download(base, NULL);
+        if (!base) {
+            mp_err(log, "Error downloading hardware screenshot frame.\n");
+            return NULL;
+        }
+    }
+
     int d_w, d_h;
-    mp_image_params_get_dsize(&image->params, &d_w, &d_h);
+    mp_image_params_get_dsize(&base->params, &d_w, &d_h);
 
     struct mp_image_params p = {
         .imgfmt = destfmt,
@@ -608,9 +618,9 @@ static struct mp_image *convert_image(struct mp_image *image, int destfmt,
         .h = d_h,
         .p_w = 1,
         .p_h = 1,
-        .color = image->params.color,
-        .repr = image->params.repr,
-        .chroma_location = image->params.chroma_location,
+        .color = base->params.color,
+        .repr = base->params.repr,
+        .chroma_location = base->params.chroma_location,
         .crop = {0, 0, d_w, d_h},
     };
     mp_image_params_guess_csp(&p);
@@ -629,12 +639,12 @@ static struct mp_image *convert_image(struct mp_image *image, int destfmt,
         mp_image_params_guess_csp(&p);
     }
 
-    if (mp_image_params_equal(&p, &image->params))
-        return mp_image_new_ref(image);
+    if (mp_image_params_equal(&p, &base->params))
+        return base == image ? mp_image_new_ref(image) : base;
 
     mp_verbose(log, "converted: %s\n", mp_image_params_to_str(&p));
 
-    struct mp_image *src = image;
+    struct mp_image *src = base;
     if (mp_image_crop_valid(&src->params) &&
         (mp_rect_w(src->params.crop) != src->w ||
          mp_rect_h(src->params.crop) != src->h))
@@ -642,6 +652,8 @@ static struct mp_image *convert_image(struct mp_image *image, int destfmt,
         src = mp_image_new_ref(src);
         if (!src) {
             mp_err(log, "mp_image_new_ref failed!\n");
+            if (base != image)
+                talloc_free(base);
             return NULL;
         }
         mp_image_crop_rc(src, src->params.crop);
@@ -650,6 +662,10 @@ static struct mp_image *convert_image(struct mp_image *image, int destfmt,
     struct mp_image *dst = mp_image_alloc(p.imgfmt, p.w, p.h);
     if (!dst) {
         mp_err(log, "Out of memory.\n");
+        if (src != base)
+            talloc_free(src);
+        if (base != image)
+            talloc_free(base);
         return NULL;
     }
     mp_image_copy_attributes(dst, src);
@@ -663,8 +679,10 @@ static struct mp_image *convert_image(struct mp_image *image, int destfmt,
     bool ok = mp_sws_scale(sws, dst, src) >= 0;
     talloc_free(sws);
 
-    if (src != image)
+    if (src != base)
         talloc_free(src);
+    if (base != image)
+        talloc_free(base);
 
     if (!ok) {
         mp_err(log, "Error when converting image.\n");
