diff --git a/source/decore/inter_pred.c b/source/decore/inter_pred.c
index 470c84c..6623a58 100644
--- a/source/decore/inter_pred.c
+++ b/source/decore/inter_pred.c
@@ -360,7 +360,64 @@ void uavs3d_always_inline mc_core_chroma(com_pic_t* ref, int i_src, pel *dst, in
     }
 }
 
-void com_mc(com_core_t *core, pel *pred_buf)
+static void uavs3d_always_inline com_mc_uni(com_core_t *core, pel *pred_buf)
+{
+    s8 *refi = core->refi;
+    s16 (*mv)[MV_D] = core->mv;
+    com_pic_t *pic = core->pic;
+    com_seqh_t *seqhdr = core->seqhdr;
+    int x = core->cu_pix_x;
+    int y = core->cu_pix_y;
+    int w = core->cu_width;
+    int h = core->cu_height;
+    com_ref_pic_t(*refp)[REFP_NUM] = core->refp;
+    int lidx_base = REFI_IS_VALID(refi[REFP_0]) ? REFP_0 : REFP_1;
+    int tree_status = core->tree_status;
+    int max_posx = seqhdr->pic_width  + 4;
+    int max_posy = seqhdr->pic_height + 4;
+    int max_val = (1 << seqhdr->bit_depth_internal) - 1;
+    int widx = core->cu_log2w - 2;
+    int i_dst;
+    int i_src = pic->stride_luma;
+    int i_srcc = pic->stride_chroma;
+    pel *dst;
+
+    if (tree_status != TREE_C) {
+        int lidx = lidx_base;
+        com_pic_t *ref_pic = refp[refi[lidx]][lidx].pic;
+        int qpel_gmv_x = (x << 2) + mv[lidx][MV_X];
+        int qpel_gmv_y = (y << 2) + mv[lidx][MV_Y];
+
+        if (!M32(core->cbfy)) {
+            i_dst = pic->stride_luma;
+            dst = pic->y + y * i_dst + x;
+        } else {
+            i_dst = w;
+            dst = pred_buf;
+        }
+        mc_core_luma(ref_pic, i_src, dst, i_dst, qpel_gmv_x, qpel_gmv_y, w, h, widx, max_posx, max_posy, max_val, 0);
+    }
+
+    if (tree_status != TREE_L) {
+        int lidx = lidx_base;
+        int blk_dim = w * h;
+        com_pic_t *ref_pic = refp[refi[lidx]][lidx].pic;
+        int qpel_gmv_x = (x << 2) + mv[lidx][MV_X];
+        int qpel_gmv_y = (y << 2) + mv[lidx][MV_Y];
+
+        if (!M16(core->cbfc)) {
+            i_dst = pic->stride_chroma;
+            dst = pic->uv + y / 2 * i_dst + x;
+        } else {
+            i_dst = w;
+            dst = pred_buf + blk_dim;
+        }
+
+        mc_core_chroma(ref_pic, i_srcc, dst, i_dst, qpel_gmv_x, qpel_gmv_y, w, h >> 1, widx, max_posx >> 1, max_posy >> 1, max_val, 0);
+    }
+}
+
+static void uavs3d_no_inline com_mc_bi(com_core_t *core, pel *pred_buf)
 {
     s8 *refi = core->refi;
     s16 (*mv)[MV_D] = core->mv;
@@ -382,14 +439,11 @@ void com_mc(com_core_t *core, pel *pred_buf)
     int i_src = pic->stride_luma;
     int i_srcc = pic->stride_chroma;
     pel *dst;
-    int bi = REFI_IS_VALID(refi[REFP_0]) &&
-             REFI_IS_VALID(refi[REFP_1]) &&
-            (refp[refi[REFP_0]][REFP_0].pic != refp[refi[REFP_1]][REFP_1].pic || M32(mv[REFP_0]) != M32(mv[REFP_1]));
-   
+
     if (tree_status != TREE_C) {
         int i, lidx = lidx_base;
 
-        for (i = 0; i <= bi; i++) {
+        for (i = 0; i < 2; i++) {
             com_pic_t *ref_pic = refp[refi[lidx]][lidx].pic;
             int qpel_gmv_x = (x << 2) + mv[lidx][MV_X];
             int qpel_gmv_y = (y << 2) + mv[lidx][MV_Y];
@@ -398,58 +452,64 @@ void com_mc(com_core_t *core, pel *pred_buf)
             if (i) {
                 i_dst = w;
                 dst = pred_snd;
-            } else if (!bi && !M32(core->cbfy)) {
-                i_dst = pic->stride_luma;
-                dst = pic->y + y * i_dst + x;
             } else {
                 i_dst = w;
                 dst = pred_buf;
             }
             mc_core_luma(ref_pic, i_src, dst, i_dst, qpel_gmv_x, qpel_gmv_y, w, h, widx, max_posx, max_posy, max_val, 0);
         }
-        if (bi) {
-            if (!M32(core->cbfy)) {
-                i_dst = pic->stride_luma;
-                uavs3d_funs_handle.avg_pel[widx](pic->y + y * i_dst + x, i_dst, pred_buf, pred_snd, w, h);
-            } else {
-                uavs3d_funs_handle.avg_pel[widx](pred_buf, w, pred_buf, pred_snd, w, h);
-            }
+        if (!M32(core->cbfy)) {
+            i_dst = pic->stride_luma;
+            uavs3d_funs_handle.avg_pel[widx](pic->y + y * i_dst + x, i_dst, pred_buf, pred_snd, w, h);
+        } else {
+            uavs3d_funs_handle.avg_pel[widx](pred_buf, w, pred_buf, pred_snd, w, h);
         }
     }
     if (tree_status != TREE_L) {
         int i, lidx = lidx_base;
         int blk_dim = w * h;
 
-        for (i = 0; i <= bi; i++) {
+        for (i = 0; i < 2; i++) {
             com_pic_t *ref_pic = refp[refi[lidx]][lidx].pic;
             int qpel_gmv_x = (x << 2) + mv[lidx][MV_X];
             int qpel_gmv_y = (y << 2) + mv[lidx][MV_Y];
-            
+
             lidx = !lidx;
             if (i) {
                 i_dst = w;
                 dst = pred_snd;
-            } else if (!bi && !M16(core->cbfc)) {
-                i_dst = pic->stride_chroma;
-                dst = pic->uv + y / 2 * i_dst + x;
             } else {
                 i_dst = w;
                 dst = pred_buf + blk_dim;
             }
-        
+
             mc_core_chroma(ref_pic, i_srcc, dst, i_dst, qpel_gmv_x, qpel_gmv_y, w, h >> 1, widx, max_posx >> 1, max_posy >> 1, max_val, 0);
         }
-        if (bi) {
-            if (!M16(core->cbfc)) {
-                i_dst = pic->stride_chroma;
-                uavs3d_funs_handle.avg_pel[widx](pic->uv + y / 2 * i_dst + x, i_dst, pred_buf + blk_dim, pred_snd, w, h >> 1);
-            } else {
-                uavs3d_funs_handle.avg_pel[widx](pred_buf + blk_dim, w, pred_buf + blk_dim, pred_snd, w, h >> 1);
-            }
+        if (!M16(core->cbfc)) {
+            i_dst = pic->stride_chroma;
+            uavs3d_funs_handle.avg_pel[widx](pic->uv + y / 2 * i_dst + x, i_dst, pred_buf + blk_dim, pred_snd, w, h >> 1);
+        } else {
+            uavs3d_funs_handle.avg_pel[widx](pred_buf + blk_dim, w, pred_buf + blk_dim, pred_snd, w, h >> 1);
         }
     }
 }
 
+void com_mc(com_core_t *core, pel *pred_buf)
+{
+    s8 *refi = core->refi;
+    s16 (*mv)[MV_D] = core->mv;
+    com_ref_pic_t(*refp)[REFP_NUM] = core->refp;
+    int bi = REFI_IS_VALID(refi[REFP_0]) &&
+             REFI_IS_VALID(refi[REFP_1]) &&
+            (refp[refi[REFP_0]][REFP_0].pic != refp[refi[REFP_1]][REFP_1].pic || M32(mv[REFP_0]) != M32(mv[REFP_1]));
+
+    if (bi) {
+        com_mc_bi(core, pred_buf);
+    } else {
+        com_mc_uni(core, pred_buf);
+    }
+}
+
 void uavs3d_always_inline com_affine_mc_luma(com_core_t *core, pel *dstl, int i_dstl, int x, int y, int cu_width, int cu_height, com_pic_t* ref_pic, int sub_blk_size, int lidx)
 {
     int w, h;
