diff --git a/source/decore/arm64/inter_pred_arm64.S b/source/decore/arm64/inter_pred_arm64.S
index 375f041..6f249b7 100644
--- a/source/decore/arm64/inter_pred_arm64.S
+++ b/source/decore/arm64/inter_pred_arm64.S
@@ -6368,7 +6368,7 @@ if_hor_ver_luma_w32_hor_10bit_loop_y:
     rshrn  v27.4h, v28.4s, #2
     rshrn2 v27.8h, v29.4s, #2
 
-    st1 {v26.8h, v27.8h}, [x17], #32
+    stp q26, q27, [x17], #32
 
     ext v19.16b, v30.16b, v31.16b, #2
     ext v20.16b, v30.16b, v31.16b, #4
@@ -6426,7 +6426,7 @@ if_hor_ver_luma_w32_hor_10bit_loop_y:
     rshrn2 v27.8h, v29.4s, #2
 
     subs w8, w8, #1
-    st1 {v26.8h, v27.8h}, [x17], #32
+    stp q26, q27, [x17], #32
     bgt if_hor_ver_luma_w32_hor_10bit_loop_y
 
     mov x17, sp
@@ -7505,7 +7505,7 @@ if_hor_ver_chroma_w16_hor_loop_y:
     mls v29.8h, v24.8h, v3.8h
 
     subs w8, w8, #1
-    st1 {v28.8h, v29.8h}, [x17], #32
+    stp q28, q29, [x17], #32
     bgt if_hor_ver_chroma_w16_hor_loop_y
 
 //--------------------------------
@@ -7614,7 +7614,7 @@ if_hor_ver_chroma_w16_hor_10bit_loop_y:
     rshrn2 v29.8h, v31.4s, #2
 
     subs w8, w8, #1
-    st1 {v28.8h, v29.8h}, [x17], #32
+    stp q28, q29, [x17], #32
     bgt if_hor_ver_chroma_w16_hor_10bit_loop_y
 
 //--------------------------------
